server/lib/api/users.js

```diff
@@ -97,3 +106,33 @@ async function resetPassword(ctx, next) {
     ctx.status = 204
   })
 }
+
+async function verifyEmail(ctx, next) {
+  const { id } = ctx.params
+  const { code } = ctx.query
+  const { email } = ctx.request.body
+
+  if (!id || !code || !isValidEmail(email)) {
+    throw new httpErrors.BadRequest('Invalid parameters')
+  }
+
+  const verifiedEmail = await useEmailVerificationCode(id, email, code)
+  if (!verifiedEmail) {
+    throw new httpErrors.BadRequest('Email verification code is invalid')
```

Comment:
You could check the error code + message contents. Or, since we know the client validates the presence of most of things, you could just always act as if this error (by "this error" I mean a 400/Bad Request) means invalid email or verification code for now.

All I mean is that the message to the client should be something like:
'The provided email or verification code is not valid. If the verification code matches the one you were emailed, it may have expired. Please request a new verification email and try again.'